compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp

```diff
@@ -243,4 +243,110 @@ verifyGPUMatmulPipeline(Operation *op,
+  };
+
+  SmallVector<int64_t, 3> expectedWorkgroupSize = {
+      scheduleAttr.getSubgroupNCount() * *subgroupSize,
+      scheduleAttr.getSubgroupMCount(), 1};
+
+  if (!llvm::all_of(llvm::zip_equal(workgroupSize, expectedWorkgroupSize),
+                    pairIsEqual)) {
+    return op->emitError("expected workgroup size to be [")
+           << expectedWorkgroupSize[0] << ", " << expectedWorkgroupSize[1]
+           << ", " << expectedWorkgroupSize[2] << "]";
+  }
+
+  auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+  if (!linalgOp) {
+    return success();
+  }
+
+  FailureOr<mlir::linalg::ConvolutionDimensions> convDims =
+      mlir::linalg::inferConvolutionDims(linalgOp);
+  if (succeeded(convDims)) {
+    // TODO: verify convolution cases
+    return success();
+  }
+
+  FailureOr<mlir::linalg::ContractionDimensions> contractDims =
+      mlir::linalg::inferContractionDims(linalgOp);
+  if (failed(contractDims)) {
+    return success();
+  }
+
+  SmallVector<int64_t> expectedTileSizes(linalgOp.getNumLoops(), 0);
+  int64_t mDim = contractDims->m.back();
+  int64_t nDim = contractDims->n.back();
+  int64_t kDim = contractDims->k.back();
+
+  // Right now we expect to tile all m, n, and k dimensions to 1 except the
+  // innermost.
+  for (int64_t m : llvm::drop_end(contractDims->m)) {
+    expectedTileSizes[m] = 1;
+  }
+  for (int64_t n : llvm::drop_end(contractDims->n)) {
+    expectedTileSizes[n] = 1;
+  }
+  for (int64_t k : llvm::drop_end(contractDims->k)) {
+    expectedTileSizes[k] = 1;
+  }
+
+  auto [mSize, nSize, kSize] = scheduleAttr.getIntrinsic().getMNKShape();
+  expectedTileSizes[mDim] = scheduleAttr.getSubgroupMCount() *
+                            scheduleAttr.getSubgroupMTileCount() * mSize;
+  expectedTileSizes[nDim] = scheduleAttr.getSubgroupNCount() *
+                            scheduleAttr.getSubgroupNTileCount() * nSize;
+  expectedTileSizes[kDim] = scheduleAttr.getSubgroupKTileCount() * kSize;
+
+  SmallVector<int64_t> tileShape =
+      loweringConfig.getTileSizeVals(kWorkgroupTileLevel);
+  if (tileShape.size() != expectedTileSizes.size()) {
+    return op->emitError("expected workgroup tile size to have ")
+           << expectedTileSizes.size() << " elements";
+  }
+  if (!llvm::all_of(llvm::zip_equal(tileShape, expectedTileSizes),
+                    pairIsEqual)) {
+    auto error = op->emitError("expected workgroup tile size to be [")
+                 << expectedTileSizes[0];
+    for (int i = 1, e = expectedTileSizes.size(); i < e; ++i) {
+      error << ", " << expectedTileSizes[i];
+    }
+    error << "]";
+    return error;
+  }
+
+  SmallVector<int64_t, 4> bounds = linalgOp.getStaticLoopRanges();
+  if (bounds[mDim] % expectedTileSizes[mDim] != 0 ||
+      bounds[nDim] % expectedTileSizes[nDim] != 0 ||
+      bounds[kDim] % expectedTileSizes[kDim] != 0) {
+    return op->emitError("workgroup tile size cannot divide problem size");
+  }
+
+  return success();
```

Comment:
Could we add a test for a skinny shape? Like 2x320x1280 with tile sizes [32, 32, 256]